### PR TITLE
Allow lock owners to update the price of keys

### LINKED ISSUE #138: Allowing Lock owners to update the price of keys
The `keyPrice` for each Lock is set upon creation of the `Lock` smart contract.
We expect most creators to actually "price" the keys to their lock based on a fiat value ($ or else). As such we should allow creators to update the price of the keys (`keyPrice`) on the locks easily at any point. This will require additions to the Lock smart contract but also changes in the web app.

### PR DESCRIPTION
**Pull request by jedahan**
This is the first commit learning how to make a save button, and toggle between save and edit states.

It is meant as a starting point for discussion of how edit functions are being passed through props, where state is changing, and figuring out the connectors and mapping.

Fixes #138

All this does right now is toggle between two icons (I reused another icon I saw in the icons folder)

This is my first time using React so I first want to be sure I am passing functions through props the right way, and I'm not sure how everything is wired up.

### CODE REVIEW

File: unlock-app/src/components/lock/Lock.js [hunk 1]
```diff
@@ -82,6 +86,7 @@ Lock.defaultProps = {
 
 const mapDispatchToProps = dispatch => ({
   purchaseKey: key => dispatch(purchaseKey(key)),
+  updateLockPrice: (lock, price) => dispatch(updateLockPrice(lock, price)),
```
Comment (jedahan) [inline unlock-app/src/components/lock/Lock.js]: I have no idea how this should actually be wired up. I just saw something close to what looked correct and put a placeholder in here for now.

File: unlock-app/src/components/lock/Lock.js [hunk 2]
```diff
@@ -73,6 +76,7 @@ Lock.propTypes = {
   purchaseKey: PropTypes.func.isRequired,
   config: UnlockPropTypes.configuration.isRequired,
   hideModal: PropTypes.func.isRequired,
+  updateLockPrice: PropTypes.func.isRequired,
```
Comment (jedahan) [inline unlock-app/src/components/lock/Lock.js]: Should this even be required?

File: unlock-app/src/middlewares/lockMiddleware.js [hunk 3]
```diff
@@ -141,6 +144,11 @@ export default function lockMiddleware({ getState, dispatch }) {
         web3Service.connect({ provider: action.provider })
       } else if (action.type === CREATE_LOCK) {
         web3Service.createLock(action.lock, getState().account)
+      } else if (action.type === UPDATE_LOCK_PRICE) {
```
Comment (jedahan) [inline unlock-app/src/middlewares/lockMiddleware.js]: How do I actually dispatch the UPDATE_LOCK_PRICE type action from the frontend?

File: unlock-app/src/middlewares/lockMiddleware.js [hunk 4]
```diff
@@ -141,6 +144,11 @@ export default function lockMiddleware({ getState, dispatch }) {
...
+        console.table([action])
+        console.log(
```
Comment (jedahan) [inline unlock-app/src/middlewares/lockMiddleware.js]: Also need validation here and in web3Service

File: unlock-app/src/components/creator/lock/LockIconBar.js [hunk 5]
```diff
@@ -38,7 +41,21 @@ export function LockIconBar({
       <IconBarContainer>
         <IconBar>
           <Buttons.Withdraw as="button" lock={lock} />
-          <Buttons.Edit as="button" />
+          {editing && (
```
Comment (jedahan) [inline unlock-app/src/components/creator/lock/LockIconBar.js]: Is this really the best way to toggle between two buttons?

File: unlock-app/src/components/interface/buttons/lock/Save.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
+import React from 'react'
+import Svg from '../../svg'
+import Button from '../Button'
+
+export const Save = ({ lock, updateLockPrice, toggleEditing, ...props }) => (
+  <Button
+    title="Save"
+    action={() => {
+      console.log('updateLockPrice(${state.something?})')
```
Comment (jedahan) [inline unlock-app/src/components/interface/buttons/lock/Save.js]: need to add validation

File: unlock-app/src/components/lock/Lock.js (hunk 1 above)
Comment (julien51) [reply unlock-app/src/components/lock/Lock.js]: This is wired up ;) Basically the function `mapDispatchToProps ` is courtesy of `react-redux` and is a way to invoked reducer actions on the front end application.

File: unlock-app/src/components/lock/Lock.js (hunk 2 above)
Comment (julien51) [reply unlock-app/src/components/lock/Lock.js]: Probably? I think you always want to have a function defined here because it will be invoked when users click on the `save` button.

File: unlock-app/src/middlewares/lockMiddleware.js (hunk 3 above)
Comment (julien51) [reply unlock-app/src/middlewares/lockMiddleware.js]: This is happening through your `updateLockPrice` in your `mapDispatchToProps` function above.

File: unlock-app/src/actions/lock.js [hunk 7]
```diff
@@ -21,6 +22,12 @@ export const updateLock = (address, update) => ({
   update,
 })
 
+export const updateLockPrice = (lock, price) => ({
```
Comment (julien51) [inline unlock-app/src/actions/lock.js]: This is good but please add a test to `actions/lock.test.js` to make sure the right action object is created.

File: unlock-app/src/components/creator/lock/LockIconBar.js (hunk 5 above)
Comment (julien51) [reply unlock-app/src/components/creator/lock/LockIconBar.js]: I think this works.

File: unlock-app/src/components/creator/lock/LockIconBar.js [hunk 8]
```diff
@@ -38,7 +41,21 @@ export function LockIconBar({
...
+            <Buttons.Save
+              as="button"
+              lock={lock}
+              toggleEditing={toggleEditing}
```
Comment (julien51) [inline unlock-app/src/components/creator/lock/LockIconBar.js]: I wonder if the `Buttons.Save` button should not have a `onClick` and `onCancel` properties which would be triggered respectively for each of `updateLockPrice` and `toggleEditing` ?

File: unlock-app/src/components/interface/buttons/lock/Save.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
+import React from 'react'
```
Comment (julien51) [inline unlock-app/src/components/interface/buttons/lock/Save.js]: Not sure shy you need this new component? Couldn't you plug that logic in the `Buttons.Save` button?

File: unlock-app/src/components/interface/buttons/lock/index.js [hunk 10]
```diff
@@ -7,6 +7,7 @@ import Upload from './Upload'
 import Etherscan from './Etherscan'
 import Preview from './Preview'
 import Copy from './Copy'
+import Save from './Save'
```
Comment (julien51) [inline unlock-app/src/components/interface/buttons/lock/index.js]: Wait... that was not included before? sorry :/ 
I can certainly pull that into its own PR so that it's less confusing. LMK.

File: unlock-app/src/components/interface/svg/Save.js [hunk 11]
```diff
@@ -0,0 +1,14 @@
+import React from 'react'
```
Comment (julien51) [inline unlock-app/src/components/interface/svg/Save.js]: Same here.

File: unlock-app/src/components/lock/Lock.js (hunk 2 above)
Comment (julien51) [reply unlock-app/src/components/lock/Lock.js]: Yes!

File: unlock-app/src/components/interface/svg/Save.js (hunk 11 above)
Comment (jedahan) [reply unlock-app/src/components/interface/svg/Save.js]: Same here regarding making the Save button a separate PR?

Or do you mean to merge components/interface/svg/Save.js with buttons/lock/Save?

File: unlock-app/src/components/interface/buttons/lock/index.js (hunk 10 above)
Comment (jedahan) [reply unlock-app/src/components/interface/buttons/lock/index.js]: Would you like me to make it a separate commit or a separate PR to add a save button?
Because we also need to choose a decent SVG and I'm not sure if you drew all these or found them someplace.

File: unlock-app/src/components/creator/lock/LockIconBar.js (hunk 8 above)
Comment (jedahan) [reply unlock-app/src/components/creator/lock/LockIconBar.js]: Thats much better naming, will change now

Comment (jedahan) [reply unlock-app/src/components/creator/lock/LockIconBar.js]: so how do we want to trigger `onCancel`? In my head, onCancel is part of whatever editing modal is higher up, like in CreatorLock/LockRow. 

So maybe when you hit escape? Or should we create another component to cancel the edit? Or should that not be solved in this particular PR but made a different issue?

Comment (jedahan) [reply unlock-app/src/components/creator/lock/LockIconBar.js]: nvm, dumb mistake. will add onCancel

File: unlock-app/src/components/interface/buttons/lock/index.js (hunk 10 above)
Comment (julien51) [reply unlock-app/src/components/interface/buttons/lock/index.js]: Wait actually you got me confused here. I think you do not need a save SVG button?

Here's what I think would work:
screenshot 2018-12-16 at 21 52 47
(please ignore the email stuff)

Basically when the user clicks on the pencil button, we toggle all fields back to form inputs (same UI/layout as when creating the lock), and the button bar is replaced with the grey area with both "Update" and "Cancel". Does it make sense?

Comment (jedahan) [reply unlock-app/src/components/interface/buttons/lock/index.js]: Yes this looks good

**Comment by julien51** [comment]
@jedahan I know it took a lot more time but I just sent a PR which handles the web3Service part of this PR.

Please let me know if you you can work on more on your end.

File: package.json [hunk 12]
```diff
@@ -29,4 +29,4 @@
     "url": "https://github.com/unlock-protocol/unlock/issues"
   },
   "homepage": "https://github.com/unlock-protocol/unlock#readme"
-}
\ No newline at end of file
+}
```
Comment (cellog) [inline package.json]: please revert all changes to package.json. You'll probably need to rebase against master.

```bash
git fetch
git rebase origin/master
git push --force
```

If you have conflicts, you'll need to resolve those. If you're uncomfortable doing rebase, you can also do a merge commit, github has a way to do a pull request to your branch from the main one which you can then merge.

File: unlock-app/package.json [hunk 13]
```diff
@@ -69,7 +69,7 @@
   "scripts": {
     "dev": "npm run setup-dev && npm run set-link && nodemon src/server.js",
     "set-link": "run-script-os",
-    "set-link:darwin:freebsd:linux:sunos": "cd src && (rm artifacts -f) && ln -s ../../smart-contracts/build artifacts",
+    "set-link:darwin:freebsd:linux:sunos": "pushd src && (test -e artifacts && rm -f artifacts) && ln -s ../../smart-contracts/build artifacts",
```
Comment (cellog) [inline unlock-app/package.json]: same comment as above, rebase will fix this

File: unlock-app/src/__tests__/components/creator/CreatorLock.test.js [hunk 14]
```diff
@@ -10,6 +10,8 @@ jest.mock('next/link', () => {
   return ({ children }) => children
 })
 
+const toggleEditing = jest.fn()
```
Comment (cellog) [inline unlock-app/src/__tests__/components/creator/CreatorLock.test.js]:
```suggestion
```

let's not do this. Since you're not testing the toggle result in this test, we don't need this

File: unlock-app/src/__tests__/components/creator/CreatorLock.test.js [hunk 15]
```diff
@@ -10,6 +10,8 @@ jest.mock('next/link', () => {
...
+
```
Comment (cellog) [inline unlock-app/src/__tests__/components/creator/CreatorLock.test.js]:
```suggestion
```

File: unlock-app/src/__tests__/components/creator/CreatorLock.test.js [hunk 16]
```diff
@@ -44,7 +46,11 @@ describe('CreatorLock', () => {
 
     let wrapper = rtl.render(
       <Provider store={store} config={config}>
-        <CreatorLock lock={lock} transaction={transaction} />
+        <CreatorLock
+          lock={lock}
+          transaction={transaction}
+          toggleEditing={toggleEditing}
```
Comment (cellog) [inline unlock-app/src/__tests__/components/creator/CreatorLock.test.js]:
```suggestion
          toggleEditing={() => {}}
```

File: unlock-app/src/__tests__/components/creator/CreatorLock.test.js [hunk 17]
```diff
@@ -81,7 +87,11 @@ describe('CreatorLock', () => {
 
     let wrapper = rtl.render(
       <Provider store={store} config={config}>
-        <CreatorLock lock={keylock} transaction={transaction} />
+        <CreatorLock
+          lock={keylock}
+          transaction={transaction}
+          toggleEditing={toggleEditing}
```
Comment (cellog) [inline unlock-app/src/__tests__/components/creator/CreatorLock.test.js]:
```suggestion
          toggleEditing={() => {}}
```

File: unlock-app/src/config.js [hunk 18]
```diff
@@ -19,13 +19,13 @@ export function getCurrentProvider(environment) {
 
   if (
     environment.web3.currentProvider.host &&
-    environment.web3.currentProvider.host.indexOf('infura') !== -1
+    environment.web3.currentProvider.host.includes('infura')
```
Comment (cellog) [inline unlock-app/src/config.js]: please revert these well-intentioned but unrelated changes (1 PR, 1 purpose)

File: unlock-app/src/middlewares/lockMiddleware.js [hunk 19]
```diff
@@ -134,6 +134,11 @@ export default function lockMiddleware({ getState, dispatch }) {
         web3Service.connect({ provider: action.provider })
       } else if (action.type === CREATE_LOCK) {
         web3Service.createLock(action.lock, getState().account)
+      } else if (action.type === UPDATE_LOCK_PRICE) {
+        const { lock } = action
+        console.table(lock)
```
Comment (cellog) [inline unlock-app/src/middlewares/lockMiddleware.js]:
```suggestion
```

File: unlock-app/src/middlewares/lockMiddleware.js [hunk 20]
```diff
@@ -134,6 +134,11 @@ export default function lockMiddleware({ getState, dispatch }) {
...
+        const { address, keyPrice } = lock
+        console.log(`web3Service.updatePrice(${address}, ${keyPrice})`)
```
Comment (cellog) [inline unlock-app/src/middlewares/lockMiddleware.js]:
```suggestion
```

File: unlock-app/src/reducers/accountReducer.js [hunk 21]
```diff
@@ -5,7 +5,7 @@ import { SET_NETWORK } from '../actions/network'
 export const initialState = null
 
 const accountReducer = (state = initialState, action) => {
-  if ([SET_PROVIDER, SET_NETWORK].indexOf(action.type) > -1) {
+  if ([SET_PROVIDER, SET_NETWORK].includes(action.type)) {
```
Comment (cellog) [inline unlock-app/src/reducers/accountReducer.js]:
```suggestion
  if ([SET_PROVIDER, SET_NETWORK].indexOf(action.type) > -1) {
```

1 PR , 1 purpose

File: unlock-app/src/reducers/errorReducer.js [hunk 22]
```diff
@@ -5,7 +5,7 @@ import { SET_NETWORK } from '../actions/network'
 export const initialState = null
 
 const errorReducer = (state = initialState, action) => {
-  if ([SET_PROVIDER, SET_NETWORK].indexOf(action.type) > -1) {
+  if ([SET_PROVIDER, SET_NETWORK].includes(action.type)) {
```
Comment (cellog) [inline unlock-app/src/reducers/errorReducer.js]:
```suggestion
  if ([SET_PROVIDER, SET_NETWORK].indexOf(action.type) > -1) {
```

1 PR , 1 purpose

File: unlock-app/src/reducers/keysReducer.js [hunk 23]
```diff
@@ -5,7 +5,7 @@ import { SET_NETWORK } from '../actions/network'
 export const initialState = {}
 
 const keysReducer = (state = initialState, action) => {
-  if ([SET_PROVIDER, SET_NETWORK].indexOf(action.type) > -1) {
+  if ([SET_PROVIDER, SET_NETWORK].includes(action.type)) {
```
Comment (cellog) [inline unlock-app/src/reducers/keysReducer.js]:
```suggestion
  if ([SET_PROVIDER, SET_NETWORK].indexOf(action.type) > -1) {
```

1 PR , 1 purpose

File: unlock-app/src/reducers/locksReducer.js [hunk 24]
```diff
@@ -11,7 +11,7 @@ import { SET_NETWORK } from '../actions/network'
 export const initialState = {}
 
 const locksReducer = (state = initialState, action) => {
-  if ([SET_PROVIDER, SET_NETWORK].indexOf(action.type) > -1) {
+  if ([SET_PROVIDER, SET_NETWORK].includes(action.type)) {
```
Comment (cellog) [inline unlock-app/src/reducers/locksReducer.js]:
```suggestion
  if ([SET_PROVIDER, SET_NETWORK].indexOf(action.type) > -1) {
```

1 PR , 1 purpose

File: unlock-app/src/reducers/modalsReducer.js [hunk 25]
```diff
@@ -5,7 +5,7 @@ import { SET_NETWORK } from '../actions/network'
 export const initialState = {}
 
 const errorReducer = (state = initialState, action) => {
-  if ([SET_PROVIDER, SET_NETWORK].indexOf(action.type) > -1) {
+  if ([SET_PROVIDER, SET_NETWORK].includes(action.type)) {
```
Comment (cellog) [inline unlock-app/src/reducers/modalsReducer.js]:
```suggestion
  if ([SET_PROVIDER, SET_NETWORK].indexOf(action.type) > -1) {
```

1 PR , 1 purpose

File: unlock-app/src/reducers/transactionsReducer.js [hunk 26]
```diff
@@ -10,7 +10,7 @@ import { SET_NETWORK } from '../actions/network'
 export const initialState = {}
 
 const transactionsReducer = (transactions = initialState, action) => {
-  if ([SET_PROVIDER, SET_NETWORK].indexOf(action.type) > -1) {
+  if ([SET_PROVIDER, SET_NETWORK].includes(action.type)) {
```
Comment (cellog) [inline unlock-app/src/reducers/transactionsReducer.js]:
```suggestion
  if ([SET_PROVIDER, SET_NETWORK].indexOf(action.type) > -1) {
```

1 PR , 1 purpose

File: unlock-app/src/registerServiceWorker.js [hunk 27]
```diff
@@ -90,7 +90,7 @@ function checkValidServiceWorker(swUrl) {
       // Ensure service worker exists, and that we really are getting a JS file.
       if (
         response.status === 404 ||
-        response.headers.get('content-type').indexOf('javascript') === -1
+        !response.headers.get('content-type').includes('javascript')
```
Comment (cellog) [inline unlock-app/src/registerServiceWorker.js]:
```suggestion
        response.headers.get('content-type').indexOf('javascript') === -1
```

1 PR , 1 purpose

**Comment by jedahan** [comment]
Todone
- [x] Reduced PR to minimal changes to implement functionality
- [x] Rebased on master
- [x] Middleware tries to update the price

Todo
- [ ] manual test
- [ ] storybook stories showing what CreatorLockForm looks like when editing is true
- [ ] unit tests for toggleEditing behavior of LockIconBar
